packages/jsii/lib/assembler.ts

```diff
@@ -417,12 +427,21 @@ export class Assembler implements Emitter {
       // Unresolvable module... We'll let tsc report this for us.
       return;
     }
-    if (resolution.resolvedModule.isExternalLibraryImport) {
+    if (
+      // We're not looking into a dependency's namespace exports, and the resolution says it's external
+      (packageRoot === this.projectInfo.projectRoot && resolution.resolvedModule.isExternalLibraryImport)
+      // Or the module resolves outside of the current dependency's tree entirely
+      || !resolution.resolvedModule.resolvedFileName.startsWith(packageRoot)
+      // Or the module is under one the current dependency's node_modules subtree
+      || resolution.resolvedModule.resolvedFileName.split(path.sep).filter(entry => entry === 'node_modules').length
+          !== packageRoot.split(path.sep).filter(entry => entry === 'node_modules').length
+    ) {
       // External re-exports are "pure-javascript" sugar; they need not be
       // represented in the jsii Assembly since the types in there will be
       // resolved through dependencies.
       return;
     }
```

Comment:
We are preventing creating a second name for the same actual type, because while TypeScript will totally follow the re-export and understand the type is just being aliased; other languages that do not feature structural type-checking will consider the two instances of the type as entirely distinct entities - that is no bueno.